include/beast/core/impl/handler_ptr.ipp

```diff
@@ -0,0 +1,126 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_IMPL_HANDLER_PTR_HPP
+#define BEAST_IMPL_HANDLER_PTR_HPP
+
+#include <boost/asio/detail/handler_alloc_helpers.hpp>
+#include <boost/assert.hpp>
+#include <memory>
+
+namespace beast {
+
+template<class T>
+template<class DeducedHandler, class... Args>
+inline
+handler_ptr<T>::P::
+P(DeducedHandler&& handler, Args&&... args)
+    : n(1)
+{
+    t = reinterpret_cast<T*>(
+        boost_asio_handler_alloc_helpers::
+            allocate(sizeof(T), handler));
+    try
+    {
+        t = new(t) T{
+            std::forward<DeducedHandler>(handler),
+                std::forward<Args>(args)...};
+    }
+    catch(...)
+    {
+        boost_asio_handler_alloc_helpers::
+            deallocate(t, sizeof(T), handler);
+        throw;
+    }
+}
+
+template<class T>
+inline
+void
+handler_ptr<T>::
+release()
+{
+    BOOST_ASSERT(p_);
+    if(--p_->n)
+        return;
+    if(p_->t)
+    {
+        // VFALCO This should be a move instead
```

Comment:
Ouch, I see.

You're probably already aware, but the Net TS replaces the allocation hooks with a single `get_associated_allocator(handler)` function. This should have the benefit that the forwarding chain only has to be ascended a single time, at operation initiation time, rather than once for every allocation and once for every deallocation. This might mitigate the problem, because it ideally would return an allocator that won't depend on any individual handler participating in the forwarding chain, save perhaps the top-most.